src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +400,122 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
-                        foreach (var diagnosticId in diagnostic.DiagnosticIds)
-                        {
-                            if (context.EnabledFixIds.IsFixIdEnabled(diagnosticId))
-                            {
-                                enabedDiagnosticSets.Add(diagnostic);
-                                break;
-                            }
-                        }
-                    }
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            // FixProjectAsync reports progress for documents within a project, so we use that information to update
+            // the solution progress.
+            var projectProgressTracker = new ProgressTracker((description, completedItems, totalItems) =>
+            {
+                progressTracker.Description = description;
 
-                    var isRemoveUnusedUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(RemoveUnusedImportsFixId);
-                    var isSortUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(SortImportsFixId);
-                    var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
-                        new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+                // The project progress tracker will report a TotalItems only up to the number of projects processed so
+                // far, but the total number of completed items will be correct. Assert these conditions in debug
+                // builds, but use bounds checks to ensure we don't throw exceptions at runtime.
+                Debug.Assert(completedItems <= progressTracker.TotalItems);
+                completedItems = Math.Min(progressTracker.CompletedItems, completedItems);
 
-                    var newDoc = await codeCleanupService.CleanupAsync(
-                        document, enabledDiagnostics, progressTracker, cancellationToken).ConfigureAwait(true);
+                Debug.Assert(completedItems - progressTracker.CompletedItems >= 0);
+                var newCompletedItems = Math.Max(0, completedItems - progressTracker.CompletedItems);
 
-                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                    return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
+                for (var i = 0; i < newCompletedItems; i++)
+                {
+                    progressTracker.ItemCompleted();
+                }
+            });
+
+            // Prepopulate the solution progress tracker with the total number of documents to process
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
```

Comment:
since this is important check split to 2 different method, might be good to extract this to its own method like CanProcessProject? and put conditions there so that later when we add new condition it get automatically picked up by both places?

Reply:
:arrow_right: Extracted helper `CanCleanupProject`